src/compiler/hxb/hxbWriter.ml

```diff
@@ -0,0 +1,1465 @@
+
+	val own_classes = new pool
+	val own_abstracts = new pool
+	val own_enums = new pool
+	val own_typedefs = new pool
+
+	val type_param_lut = new pool
+	val mutable ttp_key = None
+	val mutable type_type_parameters = new pool
+	val mutable field_type_parameters = new pool
+
+	(* Chunks *)
+
+	method start_chunk (kind : chunk_kind) =
+		(* Printf.eprintf "Writing chunk %s\n" (string_of_chunk_kind kind); *)
+		let new_chunk = new chunk kind cp in
+		DynArray.add chunks new_chunk;
+		chunk <- new_chunk
+
+	(* Basic compounds *)
+
+	method write_path (path : path) =
+		chunk#write_list (fst path) chunk#write_string;
+		chunk#write_string (snd path);
+
+	method write_full_path (pack : string list) (mname : string) (tname : string) =
+		chunk#write_list pack chunk#write_string;
+		chunk#write_string mname;
+		chunk#write_string tname;
+
+	method write_documentation (doc : doc_block) =
+		chunk#write_option doc.doc_own (fun s ->
+			chunk#write_uleb128 (docs#get s)
+		);
+		chunk#write_list doc.doc_inherited (fun s ->
+			chunk#write_uleb128 (docs#get s)
+		);
+
+	method write_pos (p : pos) =
+		chunk#write_string p.pfile;
+		chunk#write_leb128 p.pmin;
+		chunk#write_leb128 p.pmax;
+
+	method write_metadata_entry ((meta,el,p) : metadata_entry) =
+		chunk#write_string (Meta.to_string meta);
+		self#write_pos p;
+		chunk#write_list el self#write_expr;
+
+	method write_metadata ml =
+		chunk#write_list ml self#write_metadata_entry
+
+	(* References *)
+
+	method write_class_ref (c : tclass) =
+		let i = classes#get_or_add c.cl_path c in
+		(* Printf.eprintf "  Write class ref %d for %s\n" i (snd c.cl_path); *)
+		chunk#write_uleb128 i
+
+	method write_enum_ref (en : tenum) =
+		let i = enums#get_or_add en.e_path en in
+		(* Printf.eprintf "  Write enum ref %d for %s\n" i (snd en.e_path); *)
+		chunk#write_uleb128 i
+
+	method write_typedef_ref (td : tdef) =
+		let i = typedefs#get_or_add td.t_path td in
+		(* Printf.eprintf "  Write typedef ref %d for %s\n" i (s_type_path td.t_path); *)
+		chunk#write_uleb128 i
+
+	method write_abstract_ref (a : tabstract) =
+		let i = abstracts#get_or_add a.a_path a in
+		(* Printf.eprintf "  Write abstract ref %d for %s\n" i (snd a.a_path); *)
+		chunk#write_uleb128 i
+
+	method write_anon_ref (an : tanon) =
+		let pfm = Option.get (anon_id#identify true (TAnon an)) in
+		let i = anons#get_or_add pfm.pfm_path (an,ttp_key) in
+		(* Printf.eprintf "  Write anon ref %d for %s\n" i (s_type_path pfm.pfm_path); *)
+		chunk#write_uleb128 i
+
+	method write_field_ref (source : field_source) (cf : tclass_field) =
+		chunk#write_string cf.cf_name
+
+	method write_enum_field_ref ef =
+		chunk#write_string ef.ef_name
+
+	(* Type instances *)
+
+	method write_type_instance t =
+		let write_function_arg (n,o,t) =
+			chunk#write_string n;
+			chunk#write_bool o;
+			self#write_type_instance t;
+		in
+		match t with
+		| TMono r ->
+			begin match r.tm_type with
+			| None ->
+				chunk#write_byte 0
+			| Some t ->
```

Comment:
I wonder if it really makes sense to encode this case. Intuitively I'd say that we can just encode the actual type.

Reply:
I'm not sure what you mean here :thinking: 

I got that from https://github.com/HaxeFoundation/haxe/compare/development...HaxeFoundation:haxe:hxb_2020#diff-5cf7cb81443ee6a8a7338d0ffff5e7d458cac6cb763b22e318d2c0e78edf858eR246

Reply:
Yes but I don't see why we need to keep these monos around if they have an associated type. That's not super important right now though.